### PR TITLE
chore: basic create experiment endpoint [DET-4386]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

- grpc-gateway backed create experiment endpoint
- diff with the previous, manually added swagger spec 
[diff.txt](https://github.com/determined-ai/determined/files/5418986/diff.txt)

## Test Plan

- https://github.com/determined-ai/api-bindings/pull/1

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

**Comment by hamidzr** [comment]
@stoksc there are couple things I wanna chat w/ you about on this PR. let me know when it's a good time.

File: master/internal/api_experiment.go [hunk 1]
```diff
@@ -477,3 +485,62 @@ func (a *apiServer) GetExperimentCheckpoints(
...
+	}
+
+	dbExp.OwnerID = &user.ID
+	e, err := newExperiment(a.m, dbExp)
+	if err != nil {
+		return nil, errors.Wrap(err, "starting experiment")
+	}
+	a.m.system.ActorOf(actor.Addr("experiments", e.ID), e)
+
+	// DISCUSS if we could convert model.Experiment to proto experiment in memory..
```
Comment (stoksc) [inline master/internal/api_experiment.go:540]: What would be the hold ups with doing this?

**Review by stoksc — COMMENTED** [review]
This mostly LGTM. I would be in favor of this, especially as opposed the the butchery required to hack this endpoint into the swagger and correct route.

**Comment by stoksc** [comment]
@hamidzr any time is good Tuesday.

File: master/internal/api_experiment.go [hunk 2]
```diff
@@ -312,7 +323,7 @@ func (a *apiServer) KillExperiment(
 		return nil, err
 	}
 
-	addr := actor.Addr("experiments", req.Id).String()
```
Comment (stoksc) [inline master/internal/api_experiment.go:315]: I think it's fine how it was, personally. Any reason for this?

File: master/internal/api_experiment.go [hunk 3]
```diff
@@ -477,3 +488,52 @@ func (a *apiServer) GetExperimentCheckpoints(
...
+	}
+
+	if validateOnly {
+		// old api: return nil, c.NoContent(http.StatusNoContent)
+		// or we can use codes.already exists.
+		// if we don't want to respond with an error we could
+		// 1. change the response to a union that has an OK
+		// 2. convert dbExp to the proto equivalent and respond with that.
+		// 3. return an empty or null experiment
+		// return nil, status.Errorf(codes.Aborted, "experiment is valid")
```
Comment (stoksc) [inline master/internal/api_experiment.go:518]: > return nil, status.Errorf(codes.Aborted, "experiment is valid")

This is fine, assuming that maps to a 204. I think it's just a gRPC quirk that it's an "error" but as long as it returns 2XX it's fine. If someone is using the gRPC client, theyll just need to inspect the code.

File: master/internal/api_experiment.go [hunk 4]
```diff
@@ -57,16 +60,24 @@ func (a *apiServer) checkExperimentExists(id int) error {
...
+	switch err := a.m.db.QueryProto("get_experiment", exp, experimentID); {
 	case err == db.ErrNotFound:
-		return nil, status.Errorf(codes.NotFound, "experiment not found: %d", req.ExperimentId)
+		return nil, status.Errorf(codes.NotFound, "experiment not found: %d", experimentID)
 	case err != nil:
 		return nil, errors.Wrapf(err,
-			"error fetching experiment from database: %d", req.ExperimentId)
+			"error fetching experiment from database: %d", experimentID)
+	}
+	return exp, nil
```
Comment (stoksc) [inline master/internal/api_experiment.go:72]: any reason we left the config out of this?

File: master/internal/core_experiment.go [hunk 5]
```diff
@@ -338,23 +338,31 @@ func (m *Master) patchExperiment(c echo.Context) (interface{}, error) {
...
+func (m *Master) parseExperiment(body []byte) (*CreateExperimentParams, error) {
+	var params CreateExperimentParams
 	if err := json.Unmarshal(body, &params); err != nil {
-		return nil, false, errors.Wrap(err, "invalid experiment params")
+		return nil, errors.Wrap(err, "invalid experiment params")
 	}
+	return &params, nil
+}
 
+func (m *Master) postParseExperiment(params *CreateExperimentParams) (
```
Comment (stoksc) [inline master/internal/core_experiment.go:363]:
```suggestion
func (m *Master) parseCreateExperiment(params *CreateExperimentParams) (
```
maybe?

**Review by stoksc — COMMENTED** [review]
Just had some more questions to get started, but it looks really good.

File: master/internal/api_experiment.go (hunk 4 above)
Comment (hamidzr) [reply master/internal/api_experiment.go:72]: I wasn't sure if we want it the user provides the config in the same request. but maybe we add default to it?

File: master/internal/api_experiment.go (hunk 2 above)
Comment (hamidzr) [reply master/internal/api_experiment.go:315]: just remove/lower the possibility of typos when typing the address.

File: master/internal/api_experiment.go (hunk 3 above)
Comment (hamidzr) [reply master/internal/api_experiment.go:518]: hmm no this responds with 409. if the error code is in 200 range the the gateway expected to return a valid experiment of nil

https://github.com/grpc-ecosystem/grpc-gateway/blob/master/third_party/googleapis/google/rpc/code.proto

File: master/internal/api_experiment.go (hunk 4 above)
Comment (hamidzr) [reply master/internal/api_experiment.go:72]: the old api does provide it

Comment (stoksc) [reply master/internal/api_experiment.go:72]: I would just add it, personally. The user may be interested with how defaults got filled in, I think that would show?

File: proto/src/determined/api/v1/experiment.proto [hunk 6]
```diff
@@ -226,3 +227,20 @@ message GetExperimentValidationHistoryResponse {
   repeated determined.experiment.v1.ValidationHistoryEntry validation_history =
       1;
 }
+
+// Request to create a new experiment.
+message CreateExperimentRequest {
+  // Experiment context.
+  repeated determined.util.v1.File model_definition = 1;
+  // Experiment config.
+  google.protobuf.Struct config = 2;
```
Comment (stoksc) [inline proto/src/determined/api/v1/experiment.proto:236]: To match the old API and avoid any confusion:
```suggestion
  google.protobuf.Struct experiment_config = 2;
```

Comment (stoksc) [inline proto/src/determined/api/v1/experiment.proto:236]: I think this needs to be a string. The experiment config is most frequently submitted in YAML. We could use `google.protobuf.Struct` and have the client (cli, sdk, whatever) muck with it and convert it to this protobuf-struct form but I think it's better to just accept it and leave it alone, for audit purposes (people will want to see exactly what was submitted) and for simplicity (then we can just call yaml.Unmarshal and get the same ser/deser/error messages we're used to).

**Review by stoksc — APPROVED** [review]
lgtm

**Comment by stoksc** [comment]
@hamidzr how have you been testing this j/c?

**Comment by hamidzr** [comment]
> @hamidzr how have you been testing this j/c?

I tried creating an experiment with the cli, and generated python binding. (Linked in the first post) also tried forking through the generated binding.

**Comment by stoksc** [comment]
ah, ok cool. was wondering. I was just hacking the CLI up to use it.

File: proto/patches/api.json [hunk 7]
```diff
@@ -25,98 +25,6 @@
...
-          "type": "object",
-          "description": "The experiment config."
-        }
-      },
-      "description": "Response to GetExperimentRequest.",
-      "title": "GetExperimentResponse"
-    }
-  }
+  "paths": {},
+  "definitions": {}
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply proto/patches/api.json:29]: these don't do anything right now. just placeholders for the future.

[the remark this answers is not in the document]
Comment (hamidzr) [reply proto/patches/api.json:29]: this file doesn't get any schema validation so I thought it might help to leave the structure behind.

**Pull request merged**